src/support.js

```diff
@@ -302,14 +302,31 @@ function loadDynamicLibrary(lib, flags) {
     dso.module = libModule;
   }
 
+  function cleanupOnError() {
+    var lib_record = LDSO.loadedLibs[handle];
+    delete LDSO.loadedLibNames[lib_record.name];
+    delete LDSO.loadedLibs[handle];
+  }
+
   if (flags.loadAsync) {
-    return getLibModule().then(function(libModule) {
-      moduleLoaded(libModule);
-      return handle;
-    })
+    return getLibModule()
+      .catch(function (e) {
+        cleanupOnError();
+        throw e;
+      })
+      .then(function (libModule) {
+        moduleLoaded(libModule);
+        return handle;
+      })
   }
 
-  moduleLoaded(getLibModule());
+  try {
+    moduleLoaded(getLibModule());
```

Comment:
It looks like just the `getLibModule()` call can be in the try block and the `moduleLoaded()` call can be moved below?   (At least that looks like it would be in line with what the async code is doing above.